### DISCUSSION TITLE
Get state of physical inputs on a heat pump

### DISCUSSION THREAD

**Discussion by masterx1981** [emsesp/EMS-ESP32#1723] [Ideas]
Hi!
I have a Buderus WPLS 8.2 (recognized by ems esp as a Enviline, Compress 6000AW/...) , equipped with an HMC310 controller (recognized by ems esp as an RC300/RC310/...).
I'm trying to get the logic state of the inputs. Today i've played a bit with the bus, and i've discovered that when no inputs are inverted (i have available only i1 and i4) i get:
HpPower(0x048D), data: 41 00 10 08 14 00 00 00 00 00 00 00 00
If i close to gnd i1 i get:
HpPower(0x048D), data: 41 10 10 08 14 00 00 00 00 00 00 00 00
If i close i4 i get:
HpPower(0x048D), data: 41 80 10 08 14 00 00 00 00 00 00 00 00
If i close both i get:
HpPower(0x048D), data: 41 90 10 08 14 00 00 00 00 00 00 00 00

If i invert the logic of i1 (without anything closed), i get:
HpPower(0x048D), data: 41 10 10 08 14 00 00 00 00 00 00 00 00
While if i invert the logic of i4 i get:
HpPower(0x048D), data: 41 80 10 08 14 00 00 00 00 00 00 00 00

So for know the REAL state of the inputs, the information of that byte must be integrated with the info on the hpinNopt...

Now i use it as a custom entity on ha, but would be nice to have it as a string interpreted as "bit array" (like the hpinNopt).

Thanks!

**Comment by proddy** [comment]
Can you give some examples of what the string would like look? Also a nice name, and then we can code it.

**Reply by masterx1981** [reply]
From yesterday i've not seen any other value than what already discovered. My heat pump have available only i1 and i4 (i1 is connected to an on-off thermostat, and i4 for now is disconnected), but seem that some heatpump out there have available i1, i2, i3 and i4. I suspect that also the other input states are in that message/offset.
For not have an mqtt sensor for every one of them, i was thinking to create something like this (obviously read only):
![Screenshot_20240428_163239_Chrome](https://github.com/emsesp/EMS-ESP32/assets/76602320/e488b250-5bb2-4504-adc2-cbc1cae49141)
But if you consider that have a separate entity for every input is ok, the result will be the same...
For the name, hpinstate - input state (in case of single "array") or hpin1state/hpin4state-input 1 state/input 4 state (in case of separated entities) would be ok?
Thanks!

**Reply by proddy** [reply]
Ok, we'll need @MichaelDvP's brilliant mind to come up with the bit format

**Reply by MichaelDvP** [reply]
I've added the `input x state` entities to dev8, next PR.
@masterx1981 good catch.

**Comment by masterx1981** [comment]
As soon as i have some spare time i'll try it. Really thanks!
